modules/dag-l1/src/main/scala/org/tessellation/dag/l1/infrastructure/transaction/TransactionService.scala

```diff
@@ -0,0 +1,118 @@
+
+  def make[F[_]: Sync](
+    lastAccepted: MapRef[F, Address, Option[TransactionReference]],
+    waitingTransactions: MapRef[F, Address, Option[NonEmptySet[Signed[Transaction]]]]
+  ): TransactionService[F] =
+    new TransactionService[F] {
+
+      def isParentAccepted(transaction: Transaction): F[Boolean] =
+        (transaction.parent != TransactionReference.empty)
+          .guard[Option]
+          .fold(true.pure[F]) { _ =>
+            lastAccepted(transaction.source).get.map(_.contains(transaction.parent))
+          }
+
+      def getLastAcceptedReference(source: Address): F[TransactionReference] =
+        lastAccepted(source).get.map(_.getOrElse(TransactionReference.empty))
+
+      def accept(hashedTx: Hashed[Transaction]): F[Unit] = {
+        val parent = hashedTx.signed.value.parent
+        val source = hashedTx.signed.value.source
+        val reference = TransactionReference(hashedTx.hash, hashedTx.signed.value.ordinal)
+
+        lastAccepted(source)
+          .modify[Either[TransactionAcceptanceError, Unit]] { lastAccepted =>
+            if (lastAccepted.contains(parent) || (lastAccepted.isEmpty && reference.ordinal == TransactionOrdinal.first))
+              (reference.some, ().asRight)
+            else
+              (lastAccepted, ParentNotAccepted(source, lastAccepted, reference).asLeft)
+          }
+          .flatMap(_.liftTo[F])
+      }
+
+      def put(transactions: Set[Signed[Transaction]]): F[Unit] =
+        transactions
+          .groupBy(_.value.source)
+          .toList
+          .traverse {
+            case (source, txs) =>
+              waitingTransactions(source).modify { current =>
+                val updated = current match {
+                  case Some(waiting) =>
+                    NonEmptySet.fromSet(SortedSet.from(waiting.toNonEmptyList.toList ++ txs.toList))
+                  case None =>
+                    NonEmptySet.fromSet(SortedSet.from(txs.toList))
+                }
+
+                (updated, ())
+              }
+          }
+          .void
+
+      def pull(): F[Option[NonEmptyList[Signed[Transaction]]]] =
+        for {
+          lastAccepted <- lastAccepted.toMap
+          addresses <- waitingTransactions.keys
+          pulled <- addresses.traverse { address =>
+            waitingTransactions(address).modify {
+              case Some(waiting) =>
+                val lastTx = lastAccepted.getOrElse(address, TransactionReference.empty)
+                val pulledForAddress = Consecutive.take(lastTx, waiting.toNonEmptyList.toList)
+                val updatedStillWaiting =
+                  NonEmptySet.fromSet(SortedSet.from(pulledForAddress)) match {
+                    case Some(pulledTxs) => NonEmptySet.fromSet(waiting.diff(pulledTxs))
+                    case None            => waiting.some
+                  }
+                (updatedStillWaiting, pulledForAddress)
+              case None =>
+                (None, Seq.empty)
+            }
```

Comment:
I guess it would be good to handle the error here on the modify, log the error and default to `Seq.empty`. Working with `MapRef` makes pull for single address atomic and not for all the addresses. I guess logging and defaulting to `Seq.empty` is the best way to not go down the rabbit hole of rollbacking the pull for some addresses.

Reply:
Implemented the error handling as described above.